JCasC VSCode Dev Tools Case Study. (#2541)
jenkins-infra/jenkins.io: content/blog/2019/10/2019-10-04-vscode-caseStudy.adoc

---
layout: post

> title: "JCasC JSONSchema Case-Study Using VSCode"

Reviewer: ```suggestion
title: "Validating JCasC configuration files using Visual Studio Code"
```

opengraph:
  image: /images/logos/JCasC/JCasC.png
tags:
- community-bridge
- JCasC
- VSCode
- community
author: sladyn98
---

=== JSON Schema and VSCode

==== Configuration-as-code plugin

> The Configuration as Code plugin has been designed as an opinionated way to configure Jenkins based on human-readable declarative configuration files. Writing such a file should be feasible without being a Jenkins expert, just translating into code a configuration process one is used to executing in the web UI.The plugin uses a schema to verify the files being applied to the Jenkins instance.

Reviewer 2: ```suggestion
The Configuration as Code plugin has been designed as an opinionated way to configure Jenkins based on human-readable declarative configuration files. Writing such a file should be feasible without being a Jenkins expert, just exporting into configuration a configuration process one is used to executing in the web UI. The plugin uses a schema to verify the files being applied to the Jenkins instance.
```

> The Configuration as Code plugin has been designed as an opinionated way to configure Jenkins based on human-readable declarative configuration files. Writing such a file should be feasible without being a Jenkins expert, just translating into code a configuration process one is used to executing in the web UI.The plugin uses a schema to verify the files being applied to the Jenkins instance.

Reviewer: Link the plugin page ?

Author: Done :+1:

> The Configuration as Code plugin has been designed as an opinionated way to configure Jenkins based on human-readable declarative configuration files. Writing such a file should be feasible without being a Jenkins expert, just translating into code a configuration process one is used to executing in the web UI.The plugin uses a schema to verify the files being applied to the Jenkins instance.

Reviewer: One sentence per line, please. It is hard to suggest changes

Author: Done :+1:

> With the new JSON Schema being enabled developers can now test their yaml file against it. The new schema checks for the configurators, their types and to a certain extent the nesting of the configurators. VSCode allows us to test out the schema right out of the box with some modifications.

Reviewer 2: This seems too focused on implementation detail and it's not necessarily developers doing it. should be a more wider audience than that I think

Author: @timja Could you suggest something and I could get it committed ?

=== Steps to Enable the Schema Validation

> a) The first step includes installing the YAML plugin from redhat. Open up the Extension list in VSCode editor using `Ctrl + Shift + X`.

Reviewer: ```suggestion
a) The first step includes installing the YAML plugin from Red Hat. Open up the Extension list in VSCode editor using `Ctrl + Shift + X`.
```

Author: Done :+1:

> b) Select and install the YAML plugin by Red Hat. Reload the editor if necessary.

Reviewer 2: A URL link to the marketplace or github repo would be useful so users know they have got the right plugin

Author: This would be done once the project is released to the marketplace

Author: Done :)

> c) The next step includes inserting the JSON Schema inside the VSCode workspace. We would be including the schema in the `work` directory.

Reviewer 2: why in the work directory? most users won't be running jenkins via `hpi:run`

Author: Where do you want it, in the base directory ?

Reviewer 2: current directory?

Author: Done :)

d) Spin up your Jenkins server and navigate to `/configuration-as-code/jsonSchema`. 

e) Copy and paste this schema into a new file called `json.schema`.

f) In order to enable validation we need to include it in the workspace settings. In order to do that
navigate to `File` and then `Preference` and then `Settings`. Inside settings search for `json` and inside `settings.json` include the following configuration.

[source, json]

---
{
"yaml.schemas": {

>         "schema.json": "jenkins.yml"

Reviewer 2: the default configuration-as-code file is `jenkins.yaml` I'm not sure if both yml and yaml formats are picked up,
this will probably make it work for both but please test it:
```suggestion
        "schema.json": "jenkins.y[a]?ml"
```

Author: Yeah it works for me :)

    }
}
 ---

> You can specify a glob pattern as the value for schema.json which is the file name for the shema. This would apply the schema to all yaml files.

Reviewer 2: should provide an example

Author: Should I just copy the above one ?

Reviewer 2: yes

Author: Done :+1:

g) The following tasks can be done using VSCode:

  a) Auto completion (Ctrl + Space):
    Auto completes on all commands.

  b) Document Outlining (Ctrl + Shift + O):
  Provides the document outlining of all completed nodes in the file.

h) Create a new file under the work directory called `jenkins.yml`. For example consider the following contents for the file:

[source, yaml]
---
jenkins:
 systemMessage: “Hello World”
 numExecutors:2
 ---

i) The above yaml file is valid according to the schema and vscode should provide you with validation and autocompletion for the same.

For any suggestions or dicussions regarding the schema feel free to join our gitter link:https://gitter.im/jenkinsci/jcasc-devtools-project[channel].

> Issues can be posed link:https://github.com/jenkinsci/configuration-as-code-plugin/issues[here].

Reviewer 2: ```suggestion
Issues can be created on link:https://github.com/jenkinsci/configuration-as-code-plugin/issues[GitHub].
```

Author: Done :+1:

> ---

Reviewer 2: Needs file name updated for blog date, probably need to cherry-pick this commit in as well https://github.com/jenkins-infra/jenkins.io/pull/2734/commits/8b148229549a07e4717962f2ddac27c9d01c8328

Author: Is that commit already on master?

Reviewer 2: yes it is now, please update the file name to tomorrows date

layout: post
title: "JCasC JSONSchema Case-Study Using VSCode"
opengraph:
  image: /images/logos/JCasC/JCasC.png
tags:
- community-bridge
- JCasC
- VSCode
- community
author: sladyn98
---

> === JSON Schema and VSCode

Reviewer 2: Any background needed on the configuration-as-code plugin here?

Reviewer: +1. Some intro is needed

Reviewer: What is the purpose of of this header now?

Author: Got rid of it :)

With the new JSON Schema being enabled developers can now test their yaml file against it. The new schema checks for the configurators, their types and to a certain extent the nesting of the configurators. VSCode allows us to test out the schema right out of the box with some modifications.

=== Steps to Enable the Schema Validation

> a)The first step includes installing the YAML plugin from redhat. Open up the Extension list in VSCode editor using `Ctrl + Shift + X`.

Reviewer 2: ```suggestion
a) The first step includes installing the YAML plugin from redhat. Open up the Extension list in VSCode editor using `Ctrl + Shift + X`.
```

> The link:https://wiki.jenkins.io/display/JENKINS/Configuration+as+Code+Plugin[Configuration as Code plugin] has been designed as an opinionated way to configure Jenkins based on human-readable declarative configuration files. Writing such a file should be feasible without being a Jenkins expert, just translating into code a configuration process one is used to executing in the web UI. The plugin uses a schema to verify the files being applied to the Jenkins instance.

Reviewer 2: ```suggestion
The link:https://plugins.jenkins.io/configuration-as-code[Configuration as Code plugin] has been designed as an opinionated way to configure Jenkins based on human-readable declarative configuration files. Writing such a file should be feasible without being a Jenkins expert, just translating into code a configuration process one is used to executing in the web UI. The plugin uses a schema to verify the files being applied to the Jenkins instance.
```

> b)Select and install the YAML plugin by Red Hat. Reload the editor if necessary.

Reviewer 2: ``` suggestion
b)Select and install the YAML plugin by Red Hat. Reload the editor if necessary.
```

> c)The next step includes inserting the JSON Schema inside the VSCode workspace. We would be including the schema in the `work` directory.

Reviewer 2: ```suggestion
c) The next step includes inserting the JSON Schema inside the VSCode workspace. We would be including the schema in the `work` directory.
```

> With the new JSON Schema being enabled developers can now test their yaml file against it. The new schema checks for the configurators, their types and to a certain extent the nesting of the configurators. VSCode allows us to test out the schema right out of the box with some modifications.

Reviewer 2: how about?
```suggestion
With the new JSON Schema being enabled developers can now test their yaml file against it. The schema checks the `descriptors` i.e. configuration that can be applied to a plugin or Jenkins core, the correct type is used and help text is provided in some cases. VSCode allows us to test out the schema right out of the box with some modifications.
```

Author: Done :+1:

> d)Spin up your Jenkins server and navigate to `/configuration-as-code/jsonSchema`. 

Reviewer 2: ``` suggestion
d)Spin up your Jenkins server and navigate to `/configuration-as-code/jsonSchema`. 
```

> e)Copy and paste this schema into a new file called `json.schema`.

Reviewer 2: ```suggestion
e) Copy and paste this schema into a new file called `json.schema`.
```

> f)In order to enable validation we need to include it in the workspace settings. In order to do that

Reviewer 2: ```suggestion
f) In order to enable validation we need to include it in the workspace settings. In order to do that
```

navigate to `File` and then `Preference` and then `Settings`. Inside settings search for `json` and inside `settings.json` include the following configuration.

[source, json]

---
{
"yaml.schemas": {

>         "schema.json": "\*.yml"

Reviewer 2: I would default to this and say adjust to your file if needed
```suggestion
        "schema.json": "jenkins.yaml"
```

> f) In order to enable validation we need to include it in the workspace settings. In order to do that

Reviewer 2: ```suggestion
f) In order to enable validation we need to include it in the workspace settings.
```

Author: :+1:

> navigate to `File` and then `Preference` and then `Settings`. Inside settings search for `json` and inside `settings.json` include the following configuration.

Reviewer 2: ```suggestion
Navigate to `File` and then `Preference` and then `Settings`. Inside settings search for `json` and inside `settings.json` include the following configuration.
```

Author: :+1:

    }
}
 ---

You can specify a glob pattern as the value for schema.json which is the file name for the shema. This would apply the schema to all yaml files.

g)The following tasks can be done using VSCode:

  a)Auto completion (Ctrl + Space):
    Auto completes on all commands.

  b)Document Outlining (Ctrl + Shift + O):
  Provides the document outlining of all completed nodes in the file.

> f)Create a new file under the work directory called `jenkins.yml`. For example consider the following contents for the file:

Reviewer 3: ```suggestion
h) Create a new file under the work directory called `jenkins.yml`. For example consider the following contents for the file:
```

[source, yaml]
---
jenkins:
 systemMessage: “Hello World”
 numExecutors:2
 ---

> g) The above yaml file is valid according to the schema and vscode should provide you with validation and autocompletion for the same.

Reviewer 3: ```suggestion
i) The above yaml file is valid according to the schema and vscode should provide you with validation and autocompletion for the same.
```

> For any suggestions or dicussions regarding the schema feel free to join our gitter link:https://gitter.im/jenkinsci/jcasc-devtools-project[channel].

Reviewer 2: should have a link to report issues

> 

Reviewer 2: less whitespace please

>   numExecutors:2

Reviewer 2: ```suggestion
  numExecutors: 2
```

Author: :+1:

> For any suggestions or dicussions regarding the schema feel free to join our gitter link:https://gitter.im/jenkinsci/jcasc-devtools-project[channel].

Reviewer 2: ```suggestion
For any suggestions or dicussions regarding the schema feel free to join our link:https://gitter.im/jenkinsci/jcasc-devtools-project[gitter channel].
```

> Issues can be created on link:https://github.com/jenkinsci/configuration-as-code-plugin/issues[here].

Reviewer 2: using the text 'here' in a link is an accessibility anti-pattern, it gives no context to the reader on where they will be going without looking at the link
```suggestion
Issues can be created on link:https://github.com/jenkinsci/configuration-as-code-plugin/issues[GitHub].
```

Author: :+1:

---
layout: post
title: "Validating JCasC configuration files using Visual Studio Code"
opengraph:
  image: /images/logos/JCasC/JCasC.png
tags:
- community-bridge
- JCasC
- VSCode
- community
author: sladyn98
---

==== Configuration-as-code plugin

> The link:https://wiki.jenkins.io/display/JENKINS/Configuration+as+Code+Plugin[Configuration as Code plugin] has been designed as an opinionated way to configure Jenkins based on human-readable declarative configuration files. Writing such a file should be feasible without being a Jenkins expert, just translating into code a configuration process one is used to executing in the web UI. The plugin uses a schema to verify the files being applied to the Jenkins instance.

Reviewer 2: Please don't use the wiki link, use either github of 
```suggestion
The link:https://plugins.jenkins.io/configuration-as-code[Configuration as Code plugin] has been designed as an opinionated way to configure Jenkins based on human-readable declarative configuration files. Writing such a file should be feasible without being a Jenkins expert, just translating into code a configuration process one is used to executing in the web UI. The plugin uses a schema to verify the files being applied to the Jenkins instance.
```

=== Steps to Enable the Schema Validation

a) The first step includes installing the YAML plugin from Red Hat. Open up the Extension list in VSCode editor using `Ctrl + Shift + X`.

b) Select and install the link:https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml[YAML plugin] by Red Hat. Reload the editor if necessary.

c) The next step includes inserting the JSON Schema inside the VSCode workspace. We would be including the schema in the current directory.

d) Spin up your Jenkins server and navigate to `/configuration-as-code/jsonSchema`. 

e) Copy and paste this schema into a new file called `json.schema`.

f) In order to enable validation we need to include it in the workspace settings. In order to do that
navigate to `File` and then `Preference` and then `Settings`. Inside settings search for `json` and inside `settings.json` include the following configuration.

[source, json]

> 

Reviewer 2: ```suggestion
```

> ---

Reviewer 2: Something seems wrong here the dashes are showing up in example
![image](https://user-images.githubusercontent.com/21194782/67092462-28ed8d00-f1a7-11e9-9346-8ad1e2fcfc39.png)

Reviewer 2: ```suggestion
----
```

{
"yaml.schemas": {
        "schema.json": "y[a]?ml"
    }
}

>  ---

Reviewer 2: ```suggestion
 ----
```

> You can specify a glob pattern as the value for schema.json which is the file name for the shema. This would apply the schema to all yaml files. eg: [y[a]?ml]

Reviewer 2: ```suggestion
You can specify a glob pattern as the value for schema.json which is the file name for the schema. This would apply the schema to all yaml files. eg: `.*[y[a]?ml]`
```

g) The following tasks can be done using VSCode:

  a) Auto completion (Ctrl + Space):
    Auto completes on all commands.

  b) Document Outlining (Ctrl + Shift + O):
  Provides the document outlining of all completed nodes in the file.

h) Create a new file under the work directory called `jenkins.yml`. For example consider the following contents for the file:

[source, yaml]

> ---

Reviewer 2: Formatting looks really off:
![image](https://user-images.githubusercontent.com/21194782/67092518-4e7a9680-f1a7-11e9-8cb4-7db7431e2ab1.png)

Reviewer 2: ```suggestion
----
```

jenkins:
 systemMessage: “Hello World”
 numExecutors:2

>  ---

Reviewer 2: ```suggestion
 ----
```

i) The above yaml file is valid according to the schema and vscode should provide you with validation and autocompletion for the same.

=== Screenshots

> image: /images/logos/JCasC/JCasC.png

Reviewer 2: Image isn't loaded
![image](https://user-images.githubusercontent.com/21194782/67092529-5803fe80-f1a7-11e9-9503-c8e5ca150d16.png)

Reviewer: I assume the intention was to add a vscode screenshot.
I would like to note that this screenshot has not been resized to the blogpost layout, and that the screenshot was not cropped. 50% of the screenshot in the bottom does not contain useful info AFAICT

> image: /images/logos/JCasC/JCasC.png

Reviewer 2: ```suggestion
image:/images/projects/jcasc/vscode.png
```
